src/services/transaction_watcher_service.ts

```diff
@@ -0,0 +1,116 @@
+        //             expiresAt: LessThanOrEqual(now),
+        //         },
+        //     ],
+        // });
+        const now = new Date();
+        const isExpired = tx.expectedAt <= now;
+        try {
+            const txInBlockchain = await this._web3Wrapper.getTransactionByHashAsync(tx.hash);
+            if (txInBlockchain !== undefined && txInBlockchain !== null && txInBlockchain.hash !== undefined) {
+                if (txInBlockchain.blockNumber !== null) {
+                    logger.info(
+                        `a transaction with a ${tx.status} status is already on the blockchain, updating status to confirmed`,
+                        { hash: txInBlockchain.hash },
+                    );
+
+                    tx.status = TransactionStates.Confirmed;
+                    return this._saveTransactionAsync(tx);
+                } else if (!isExpired && tx.status !== TransactionStates.Mempool) {
+                    logger.info(`a transaction with a ${tx.status} status is pending, updating status to mempool`, {
+                        hash: txInBlockchain.hash,
+                    });
+                    tx.status = TransactionStates.Mempool;
+                    return this._saveTransactionAsync(tx);
+                } else if (isExpired) {
+                    tx.status = TransactionStates.Stuck;
+                    return this._saveTransactionAsync(tx);
+                }
+            }
+        } catch (err) {
+            // TODO(oskar) this throws a type error when calling a tx that was
+            // already dropped. Can we catch this specific error?
+            // TypeError: Cannot read property 'blockNumber' of null
+            // at Object.unmarshalTransaction (/Users/overmorrow/projects/0x/0x-api/node_modules/@0x/web3-wrapper/src/marshaller.ts:85:32)
+            // at Web3Wrapper.<anonymous> (/Users/overmorrow/projects/0x/0x-api/node_modules/@0x/web3-wrapper/src/web3_wrapper.ts:253:40)
+            // at step (/Users/overmorrow/projects/0x/0x-api/node_modules/@0x/web3-wrapper/lib/src/web3_wrapper.js:43:23)
+            // at Object.next (/Users/overmorrow/projects/0x/0x-api/node_modules/@0x/web3-wrapper/lib/src/web3_wrapper.js:24:53)
+            // at fulfilled (/Users/overmorrow/projects/0x/0x-api/node_modules/@0x/web3-wrapper/lib/src/web3_wrapper.js:15:58)
+            // at process._tickCallback (internal/process/next_tick.js:68:7)
+            if (isExpired) {
+                tx.status = TransactionStates.Dropped;
+                return this._saveTransactionAsync(tx);
+            }
+        }
+        return tx;
+    }
+    private async _saveTransactionAsync(tx: TransactionEntity): Promise<TransactionEntity> {
+        await this._transactionRepository.manager.transaction(async transactionalEntityManager => {
+            const repo = transactionalEntityManager.getRepository(TransactionEntity);
+            await repo.save(tx);
+        });
```

Comment:
This doesn't do much. Given that TransactionWatcher is a singleton instance, there is only one process attempting to update a given txn at any given time. We therefore don't need a txn here.

Reply:
There are two processes, one is a `SignerService` and `TransactionWatcher` service, and it could be the case that `Signer` will update the same transaction with `submitted` status, however we could argue that `TransactionWatcher` update should take precedence (since we do check the status in `SignerService`).

Reply:
Changed, so that `TransactionWatcher` saves without a DB transaction.